In [1]:
#lets start by reading in the csv file and printing out the first few lines
import pandas as pd

df = pd.read_csv('insurance_claims.csv')


df.head (5)


months_as_customer  age  policy_number policy_bind_date policy_state  \
0                 328   48         521585       2014-10-17           OH   
1                 228   42         342868       2006-06-27           IN   
2                 134   29         687698       2000-09-06           OH   
3                 256   41         227811       1990-05-25           IL   
4                 228   44         367455       2014-06-06           IL   

  policy_csl  policy_deductable  policy_annual_premium  umbrella_limit  \
0    250/500               1000                1406.91               0   
1    250/500               2000                1197.22         5000000   
2    100/300               2000                1413.14         5000000   
3    250/500               2000                1415.74         6000000   
4   500/1000               1000                1583.91         6000000   

   insured_zip  ... police_report_available total_claim_amount injury_claim  \
0       466132  ...                     YES              71610         6510   
1       468176  ...                       ?               5070          780   
2       430632  ...                      NO              34650         7700   
3       608117  ...                      NO              63400         6340   
4       610706  ...                      NO               6500         1300   

  property_claim vehicle_claim  auto_make  auto_model auto_year  \
0          13020         52080       Saab         92x      2004   
1            780          3510   Mercedes        E400      2007   
2           3850         23100      Dodge         RAM      2007   
3           6340         50720  Chevrolet       Tahoe      2014   
4            650          4550     Accura         RSX      2009   

  fraud_reported _c39  
0              Y  NaN  
1              Y  NaN  
2              N  NaN  
3              Y  NaN  
4              N  NaN  

[5 rows x 40 columns]

In [2]:
#lets take a look at the number of rows and columns in the dataframe
df.shape



(1000, 40)

In [3]:
df.columns

Index(['months_as_customer', 'age', 'policy_number', 'policy_bind_date',
       'policy_state', 'policy_csl', 'policy_deductable',
       'policy_annual_premium', 'umbrella_limit', 'insured_zip', 'insured_sex',
       'insured_education_level', 'insured_occupation', 'insured_hobbies',
       'insured_relationship', 'capital-gains', 'capital-loss',
       'incident_date', 'incident_type', 'collision_type', 'incident_severity',
       'authorities_contacted', 'incident_state', 'incident_city',
       'incident_location', 'incident_hour_of_the_day',
       'number_of_vehicles_involved', 'property_damage', 'bodily_injuries',
       'witnesses', 'police_report_available', 'total_claim_amount',
       'injury_claim', 'property_claim', 'vehicle_claim', 'auto_make',
       'auto_model', 'auto_year', 'fraud_reported', '_c39'],
      dtype='object')

# 📊 Insurance Claims Dataset: Field Explanations

"""
🔹 Customer & Policy Information
- months_as_customer: Number of months the person has been a customer
- age: Age of the insured person
- policy_number: Unique identifier for the policy
- policy_bind_date: Date when the policy was issued/bound
- policy_state: State where the policy was issued
- policy_csl: Combined single limit of liability coverage
- policy_deductable: Deductible amount for the policy
- policy_annual_premium: Annual premium paid for the policy
- umbrella_limit: Maximum coverage limit of the umbrella policy

🔹 Insured (Customer) Demographics
- insured_zip: ZIP code of the insured
- insured_sex: Gender of the insured
- insured_education_level: Highest education level of the insured
- insured_occupation: Occupation of the insured
- insured_hobbies: Hobbies of the insured
- insured_relationship: Relationship status of the insured

🔹 Financial Information
- capital-gains: Capital gains reported by the insured
- capital-loss: Capital losses reported by the insured
Note:

 When your car is damaged or totaled, its value often drops.
 If the car’s market value after the accident is less than what you originally paid, that’s an economic loss (a capital loss in broader financial language).

Example: You bought a car for $25,000, and before the accident it was worth $15,000. After a collision, insurance values it at $10,000 and pays you that amount.

🔹 Incident Details
- incident_date: Date of the reported incident
- incident_type: Type of incident (e.g., theft, collision, fire)
- collision_type: Type of collision (if applicable)
- incident_severity: Severity of the incident (e.g., minor, major, total loss)
- authorities_contacted: Which authorities were contacted (police, fire dept.)
- incident_state: State where the incident occurred
- incident_city: City where the incident occurred
- incident_location: Specific location/address of the incident
- incident_hour_of_the_day: Hour of the day when the incident occurred

🔹 Incident Outcomes
- number_of_vehicles_involved: How many vehicles were part of the incident
- property_damage: Whether there was property damage (Y/N/Unknown)
- bodily_injuries: Number of bodily injuries reported
- witnesses: Number of witnesses present
- police_report_available: Whether a police report was available (Y/N/Unknown)

🔹 Claim Information
- total_claim_amount: Total claim amount requested/paid
- injury_claim: Claim amount related to injuries
- property_claim: Claim amount related to property damage
- vehicle_claim: Claim amount related to vehicle damage

🔹 Vehicle Information
- auto_make: Make of the insured’s vehicle
- auto_model: Model of the insured’s vehicle
- auto_year: Year of the insured’s vehicle

🔹 Fraud Indicator
- fraud_reported: Whether the claim was flagged/reported as fraud (Y/N)

🔹 Miscellaneous
- _c39: Extra/unidentified column (may be an artifact from CSV export)
"""


In [4]:
#lets get a few summary statistics of the data
df.describe()

months_as_customer          age  policy_number  policy_deductable  \
count         1000.000000  1000.000000    1000.000000        1000.000000   
mean           203.954000    38.948000  546238.648000        1136.000000   
std            115.113174     9.140287  257063.005276         611.864673   
min              0.000000    19.000000  100804.000000         500.000000   
25%            115.750000    32.000000  335980.250000         500.000000   
50%            199.500000    38.000000  533135.000000        1000.000000   
75%            276.250000    44.000000  759099.750000        2000.000000   
max            479.000000    64.000000  999435.000000        2000.000000   

       policy_annual_premium  umbrella_limit    insured_zip  capital-gains  \
count            1000.000000    1.000000e+03    1000.000000    1000.000000   
mean             1256.406150    1.101000e+06  501214.488000   25126.100000   
std               244.167395    2.297407e+06   71701.610941   27872.187708   
min               433.330000   -1.000000e+06  430104.000000       0.000000   
25%              1089.607500    0.000000e+00  448404.500000       0.000000   
50%              1257.200000    0.000000e+00  466445.500000       0.000000   
75%              1415.695000    0.000000e+00  603251.000000   51025.000000   
max              2047.590000    1.000000e+07  620962.000000  100500.000000   

        capital-loss  incident_hour_of_the_day  number_of_vehicles_involved  \
count    1000.000000               1000.000000                   1000.00000   
mean   -26793.700000                 11.644000                      1.83900   
std     28104.096686                  6.951373                      1.01888   
min   -111100.000000                  0.000000                      1.00000   
25%    -51500.000000                  6.000000                      1.00000   
50%    -23250.000000                 12.000000                      1.00000   
75%         0.000000                 17.000000                      3.00000   
max         0.000000                 23.000000                      4.00000   

       bodily_injuries    witnesses  total_claim_amount  injury_claim  \
count      1000.000000  1000.000000          1000.00000   1000.000000   
mean          0.992000     1.487000         52761.94000   7433.420000   
std           0.820127     1.111335         26401.53319   4880.951853   
min           0.000000     0.000000           100.00000      0.000000   
25%           0.000000     1.000000         41812.50000   4295.000000   
50%           1.000000     1.000000         58055.00000   6775.000000   
75%           2.000000     2.000000         70592.50000  11305.000000   
max           2.000000     3.000000        114920.00000  21450.000000   

       property_claim  vehicle_claim    auto_year  _c39  
count     1000.000000    1000.000000  1000.000000   0.0  
mean      7399.570000   37928.950000  2005.103000   NaN  
std       4824.726179   18886.252893     6.015861   NaN  
min          0.000000      70.000000  1995.000000   NaN  
25%       4445.000000   30292.500000  2000.000000   NaN  
50%       6750.000000   42100.000000  2005.000000   NaN  
75%      10885.000000   50822.500000  2010.000000   NaN  
max      23670.000000   79560.000000  2015.000000   NaN

# 📊 Insurance Claims — Summary Statistics (Numeric Features)

Here are the descriptive statistics from the dataset (n = 1000 claims):

---

## 📅 Customer & Policy
| Variable | Count | Mean | Std | Min | 25% | 50% | 75% | Max |
|---|---:|---:|---:|---:|---:|---:|---:|---:|
| `months_as_customer` | 1000 | 203.95 | 115.11 | 0 | 115.75 | 199.5 | 276.25 | 479 |
| `age` | 1000 | 38.95 | 9.14 | 19 | 32 | 38 | 44 | 64 |
| `policy_number` | 1000 | 546,238.65 | 257,063.01 | 100,804 | 335,980 | 533,135 | 759,100 | 999,435 |
| `policy_deductable` | 1000 | 1,136.00 | 611.86 | 500 | 500 | 1000 | 2000 | 2000 |
| `policy_annual_premium` | 1000 | 1,256.41 | 244.17 | 433.33 | 1089.61 | 1257.20 | 1415.70 | 2047.59 |
| `umbrella_limit` | 1000 | 1.10e6 | 2.30e6 | -1.00e6 | 0 | 0 | 0 | 1.00e7 |

---

## 🏦 Financials
| Variable | Count | Mean | Std | Min | 25% | 50% | 75% | Max |
|---|---:|---:|---:|---:|---:|---:|---:|---:|
| `insured_zip` | 1000 | 501,214.49 | 71,701.61 | 430,104 | 448,405 | 466,446 | 603,251 | 620,962 |
| `capital-gains` | 1000 | 25,126.10 | 27,872.19 | 0 | 0 | 0 | 51,025 | 100,500 |
| `capital-loss` | 1000 | -26,793.70 | 28,104.10 | -111,100 | -51,500 | -23,250 | 0 | 0 |

---

## 🚨 Incident Details
| Variable | Count | Mean | Std | Min | 25% | 50% | 75% | Max |
|---|---:|---:|---:|---:|---:|---:|---:|---:|
| `incident_hour_of_the_day` | 1000 | 11.64 | 6.95 | 0 | 6 | 12 | 17 | 23 |
| `number_of_vehicles_involved` | 1000 | 1.84 | 1.02 | 1 | 1 | 1 | 3 | 4 |
| `bodily_injuries` | 1000 | 0.99 | 0.82 | 0 | 0 | 1 | 2 | 2 |
| `witnesses` | 1000 | 1.49 | 1.11 | 0 | 1 | 1 | 2 | 3 |

---

## 💰 Claims
| Variable | Count | Mean | Std | Min | 25% | 50% | 75% | Max |
|---|---:|---:|---:|---:|---:|---:|---:|---:|
| `total_claim_amount` | 1000 | 52,761.94 | 26,401.53 | 100 | 41,813 | 58,055 | 70,593 | 114,920 |
| `injury_claim` | 1000 | 7,433.42 | 4,880.95 | 0 | 4,295 | 6,775 | 11,305 | 21,450 |
| `property_claim` | 1000 | 7,399.57 | 4,824.73 | 0 | 4,445 | 6,750 | 10,885 | 23,670 |
| `vehicle_claim` | 1000 | 37,928.95 | 18,886.25 | 70 | 30,293 | 42,100 | 50,823 | 79,560 |

---

## 🚗 Vehicle
| Variable | Count | Mean | Std | Min | 25% | 50% | 75% | Max |
|---|---:|---:|---:|---:|---:|---:|---:|---:|
| `auto_year` | 1000 | 2005.10 | 6.02 | 1995 | 2000 | 2005 | 2010 | 2015 |

---

### ⚠️ Notes
- `umbrella_limit` has extreme outliers (min = -1,000,000, max = 10,000,000).  
- `capital-gains` and `capital-loss` are highly skewed (lots of 0s).  
- Claims data (`injury_claim`, `property_claim`, `vehicle_claim`) have heavy right tails — large maximums compared to median.  
- `_c39` column contains no values → candidate for drop.  


In [5]:
#Lets start by dropping any empty columns
import pandas as pd
import numpy as np

#lets drop the _c39 column as it is completely empty
df = df.drop(columns=['_c39'])

# umbrella_limit: inspect and cap / flag
print(df['umbrella_limit'].describe())

count    1.000000e+03
mean     1.101000e+06
std      2.297407e+06
min     -1.000000e+06
25%      0.000000e+00
50%      0.000000e+00
75%      0.000000e+00
max      1.000000e+07
Name: umbrella_limit, dtype: float64


In [6]:
#lets further inspect the umbrella_limit column
print(df['umbrella_limit'].value_counts())



umbrella_limit
 0           798
 6000000      57
 5000000      46
 4000000      39
 7000000      29
 3000000      12
 8000000       8
 9000000       5
 2000000       3
 10000000      2
-1000000       1
Name: count, dtype: int64



Umbrella insurance provides EXTRA liability coverage
on top of existing policies (like auto or homeowners).
It "kicks in" when those underlying policy limits are reached.

🔹 Example:
- Auto insurance liability limit: $250,000
- Accident causes damages: $1,000,000

Auto insurance pays first $250,000.
Umbrella policy pays the remainder up to its umbrella limit.

So, if the umbrella limit is $1,000,000:
- Auto covers: $250,000
- Umbrella covers: $750,000
- Total covered: $1,000,000
- You’re protected from paying out of pocket.

798/1000 rows have umbrella_limit = 0 → most policies have no umbrella coverage.

Many repeated large values (2–10M) suggest a set of standard coverage tiers, not continuous measurements.

One row has -1,000,000 → looks like a sentinel / data error (treat as missing).

Mean and std are dominated by the few large tiers, so raw mean is not representative.

In [7]:
#looks like one record has a negative value, which is impossible
#lets replace it with a NaN value
# ensure numeric (coerce any stray strings to NaN)
df['umbrella_limit'] = pd.to_numeric(df['umbrella_limit'], errors='coerce')

# replace negative sentinel values with NaN (keep zero as valid "no coverage")
df.loc[df['umbrella_limit'] < 0, 'umbrella_limit'] = np.nan


# show the rows where umbrella_limit is NaN (full rows)
print("Policy number(s) with umbrella_limit = NaN:")
print(df.loc[df['umbrella_limit'].isna(), 'policy_number'].tolist())


Policy number(s) with umbrella_limit = NaN:
[526039]


For policy number 526039, instead of having a negative value, which is not possible for Umbrella Coverage, it now has the value NaN in that column


In [8]:
#Now lets take a look at the capital-gains column
print(df['capital-gains'].describe())



count      1000.000000
mean      25126.100000
std       27872.187708
min           0.000000
25%           0.000000
50%           0.000000
75%       51025.000000
max      100500.000000
Name: capital-gains, dtype: float64


In [9]:
#looks like there are a lot of 0 values, which is possible
#lets take a look at the value counts
print(df['capital-gains'].value_counts())

#Lots of 0 values, which is possible
#lets check for negative values, which are not possible
print(df.loc[df['capital-gains'] < 0])


#No negative values found, so we are good here

#next question is what to do with the NaN values
#lets see how many NaN values there are
print("Number of NaN values in capital-gains column:", df['capital-gains'].isna().sum())


#There are no NaN values in this column, so we are good here, however, the data is highly skewed with a few very high values
#we will leave it as is for now, but we may want to consider capping the high values later

capital-gains
0        508
46300      5
51500      4
68500      4
55600      3
        ... 
36700      1
54900      1
69200      1
48800      1
50300      1
Name: count, Length: 338, dtype: int64
Empty DataFrame
Columns: [months_as_customer, age, policy_number, policy_bind_date, policy_state, policy_csl, policy_deductable, policy_annual_premium, umbrella_limit, insured_zip, insured_sex, insured_education_level, insured_occupation, insured_hobbies, insured_relationship, capital-gains, capital-loss, incident_date, incident_type, collision_type, incident_severity, authorities_contacted, incident_state, incident_city, incident_location, incident_hour_of_the_day, number_of_vehicles_involved, property_damage, bodily_injuries, witnesses, police_report_available, total_claim_amount, injury_claim, property_claim, vehicle_claim, auto_make, auto_model, auto_year, fraud_reported]
Index: []

[0 rows x 39 columns]
Number of NaN values in capital-gains column: 0


# 💰 Capital Gains Column Analysis

The 'capital-gains' column represents reported capital gains
associated with the insured customer.

📊 Distribution Summary:
- Most customers report **0** capital gains (508 occurrences).
- A small number of customers report positive gains at various levels:
    • 46,300 → 5 customers
    • 51,500 → 4 customers
    • 68,500 → 4 customers
    • 55,600 → 3 customers
    • ... and many other values with only 1 occurrence.
- In total, there are **338 unique values** of capital gains.

🔎 Data Quality:
- No missing values (NaN count = 0).
- Skewed distribution (heavily concentrated at 0).
- Long-tail effect: many unique positive values but low frequency.

✅ Implications for Analysis:
- Most insured individuals don’t report capital gains.
- Customers with large gains are rare but may be important outliers.
- This feature could have predictive power in fraud detection:
    • A sudden spike in unusual gains could correlate with suspicious activity.
    • Encoding should consider the extreme imbalance (0 vs. non-0).


In [10]:
#Now lets take a look at the capital-loss column
print(df['capital-loss'].describe())


count      1000.000000
mean     -26793.700000
std       28104.096686
min     -111100.000000
25%      -51500.000000
50%      -23250.000000
75%           0.000000
max           0.000000
Name: capital-loss, dtype: float64


# 📊 Capital-Loss Column Analysis (Vehicle Insurance Claims Dataset)

# Summary Statistics
 count    → 1000 (all rows have values)

 mean     → -26,793 (average claim shows ~27K in losses)

 std      → 28,104 (very high variability in losses)

 min      → -111,100 (largest recorded loss)

 25%      → -51,500 (25% of claims lost more than 51.5K)

 50%      → -23,250 (median loss ~23K)

 75%      → 0 (25% of claims report no capital loss)

 max      → 0 (no positive values; only losses or none)

# ✅ Interpretation:
- Values are strictly negative or zero.
- Negative values = reduction in value of the vehicle/property not offset by insurance.
- Zeros = either fully covered claims or cases where no capital loss applied.
- The distribution is skewed: many claims at 0, some with very large negative losses.
- Extreme negative values likely correspond to severe accidents/total losses (e.g., expensive vehicles or fleets). These exteremes will likely be a good idea for me to take a look at further in the EDA analysis


In [11]:
# Replace "?" with "Unknown" in selected columns
cols_to_fix = ['collision_type', 'property_damage', 'police_report_available']
df[cols_to_fix] = df[cols_to_fix].replace("?", "Unknown")

# Check the fix
print(df[cols_to_fix].head())

    collision_type property_damage police_report_available
0   Side Collision             YES                     YES
1          Unknown         Unknown                 Unknown
2   Rear Collision              NO                      NO
3  Front Collision         Unknown                      NO
4          Unknown              NO                      NO


### Replacing "?" with "Unknown"

In several columns, missing or unclear values are represented with `"?"`.  
To improve readability and ensure consistent categorical handling, we replace all `"?"` entries with `"Unknown"` in the following columns:

- **collision_type**  
- **property_damage**  
- **police_report_available**  

This makes the dataset cleaner and avoids confusion during analysis or visualization.


In [12]:
# Split 'Policy Csl' into two new columns
df[['Policy_Csl_Person', 'Policy_Csl_Accident']] = df['policy_csl'].str.split('/', expand=True)

# Convert to numeric for analysis
df['Policy_Csl_Person'] = pd.to_numeric(df['Policy_Csl_Person'])
df['Policy_Csl_Accident'] = pd.to_numeric(df['Policy_Csl_Accident'])

# Preview
print(df[['policy_csl', 'Policy_Csl_Person', 'Policy_Csl_Accident']].head())

#lets drop the original policy_csl column
df = df.drop(columns=['policy_csl'])


  policy_csl  Policy_Csl_Person  Policy_Csl_Accident
0    250/500                250                  500
1    250/500                250                  500
2    100/300                100                  300
3    250/500                250                  500
4   500/1000                500                 1000


### Splitting the 'Policy Csl' Column

The `policy_csl` column contains compound values (e.g., `250/500`) that represent coverage limits.  
To make analysis easier, we split this column into two new numeric fields:  

- **Policy_Csl_Person** → coverage per person  
- **Policy_Csl_Accident** → coverage per accident  

After splitting, we drop the original `policy_csl` column since it is redundant.


In [13]:
#Lets first see if we see any duplicate rows
print("Number of duplicate rows:", df.duplicated().sum())

#There are no duplicate rows, so we are good here

#Next, lets check for any missing values in the dataframe
print("Number of missing values in each column:")
print(df.isna().sum())



Number of duplicate rows: 0
Number of missing values in each column:
months_as_customer              0
age                             0
policy_number                   0
policy_bind_date                0
policy_state                    0
policy_deductable               0
policy_annual_premium           0
umbrella_limit                  1
insured_zip                     0
insured_sex                     0
insured_education_level         0
insured_occupation              0
insured_hobbies                 0
insured_relationship            0
capital-gains                   0
capital-loss                    0
incident_date                   0
incident_type                   0
collision_type                  0
incident_severity               0
authorities_contacted          91
incident_state                  0
incident_city                   0
incident_location               0
incident_hour_of_the_day        0
number_of_vehicles_involved     0
property_damage                 0
bodily_injuri

In [14]:
#It seems that authorities_contacted field has a lot of missing values
#lets take a look at the value counts for this column
print(df['authorities_contacted'].value_counts(dropna=False))   

authorities_contacted
Police       292
Fire         223
Other        198
Ambulance    196
NaN           91
Name: count, dtype: int64


#It seems that most of the values are NaN, which is possible if no authorities were contacted
#we will leave it as is for now

# 🚓 Authorities Contacted Column

"""
It seems that most of the values in the 'authorities_contacted' column are NaN.  
This makes sense, as many incidents may not have required contacting any authorities.  

✅ Decision:
We will leave the NaN values as they are for now, since:
- NaN here may carry meaningful information (e.g., "no authorities contacted").
- Dropping or imputing could remove that signal. I'll need to look further into whether NaN is the same as no authorities contacted. 
- At this time, based off of analysis, I am going to treat NaN as no authorities contacted, and change the field to no authorities contacted
"""


In [15]:
#For authorities_contacted, we have many NaN values, I'll treat NaN as no authorities contacted and change the field to no authorities contacted

df['authorities_contacted'] = df ['authorities_contacted'].fillna('No Authorities Contacted')

#lets see if that worked
print(df['authorities_contacted'].value_counts(dropna=False))

authorities_contacted
Police                      292
Fire                        223
Other                       198
Ambulance                   196
No Authorities Contacted     91
Name: count, dtype: int64


In [16]:
#Lets now normalize the text fields by making them all lowercase
df.columns = df.columns.str.strip().str.lower().str.replace('-', '_').str.replace(' ', '_')

print(df.columns)


Index(['months_as_customer', 'age', 'policy_number', 'policy_bind_date',
       'policy_state', 'policy_deductable', 'policy_annual_premium',
       'umbrella_limit', 'insured_zip', 'insured_sex',
       'insured_education_level', 'insured_occupation', 'insured_hobbies',
       'insured_relationship', 'capital_gains', 'capital_loss',
       'incident_date', 'incident_type', 'collision_type', 'incident_severity',
       'authorities_contacted', 'incident_state', 'incident_city',
       'incident_location', 'incident_hour_of_the_day',
       'number_of_vehicles_involved', 'property_damage', 'bodily_injuries',
       'witnesses', 'police_report_available', 'total_claim_amount',
       'injury_claim', 'property_claim', 'vehicle_claim', 'auto_make',
       'auto_model', 'auto_year', 'fraud_reported', 'policy_csl_person',
       'policy_csl_accident'],
      dtype='object')


In [17]:
#now lets standarize common categorical columns

cat_map = {
    'y': 'Yes', 'yes': 'Yes', 'Y': 'Yes', 'YES': 'Yes',
    'n': 'No', 'no': 'No', 'N': 'No', 'NO': 'No',
    'Unknown': 'Unknown', '?': 'Unknown',
    'none': 'None', 'None': 'None', 'NONE': 'None',

}

# Apply mapping to every value in every column that is string/object type
for col in df.select_dtypes(include="object"):
    df[col] = df[col].replace(cat_map)


#now lets start extracting date features
# Convert 'incident_date' to datetime
df['incident_date'] = pd.to_datetime(df['incident_date'], errors = 'coerce')

# Extract date features
df['incident_year'] = df['incident_date'].dt.year
df['incident_month'] = df['incident_date'].dt.month
df['incident_day'] = df['incident_date'].dt.day
df['incident_dayofweek'] = df['incident_date'].dt.dayofweek # Monday=0, Sunday=6

#Nows lets preview the new columns
print(df[['incident_date', 'incident_year', 'incident_month', 'incident_day', 'incident_dayofweek']].head())

#remove the original incident_date column
df = df.drop(columns=['incident_date'])

print(df[['incident_year', 'incident_month', 'incident_day', 'incident_dayofweek']].head())

  incident_date  incident_year  incident_month  incident_day  \
0    2015-01-25           2015               1            25   
1    2015-01-21           2015               1            21   
2    2015-02-22           2015               2            22   
3    2015-01-10           2015               1            10   
4    2015-02-17           2015               2            17   

   incident_dayofweek  
0                   6  
1                   2  
2                   6  
3                   5  
4                   1  
   incident_year  incident_month  incident_day  incident_dayofweek
0           2015               1            25                   6
1           2015               1            21                   2
2           2015               2            22                   6
3           2015               1            10                   5
4           2015               2            17                   1


In [18]:
#ensure all numeric columns are of numeric type
numeric_cols = df.select_dtypes(include=['int64', 'float64']).columns
print(numeric_cols)

#lets check if any numeric columns are of object type
object_cols = df.select_dtypes(include=['object']).columns
print(object_cols)

#Any overlap?   
overlap = set(numeric_cols).intersection(set(object_cols))
print("Overlap between numeric and object columns:", overlap)

Index(['months_as_customer', 'age', 'policy_number', 'policy_deductable',
       'policy_annual_premium', 'umbrella_limit', 'insured_zip',
       'capital_gains', 'capital_loss', 'incident_hour_of_the_day',
       'number_of_vehicles_involved', 'bodily_injuries', 'witnesses',
       'total_claim_amount', 'injury_claim', 'property_claim', 'vehicle_claim',
       'auto_year', 'policy_csl_person', 'policy_csl_accident'],
      dtype='object')
Index(['policy_bind_date', 'policy_state', 'insured_sex',
       'insured_education_level', 'insured_occupation', 'insured_hobbies',
       'insured_relationship', 'incident_type', 'collision_type',
       'incident_severity', 'authorities_contacted', 'incident_state',
       'incident_city', 'incident_location', 'property_damage',
       'police_report_available', 'auto_make', 'auto_model', 'fraud_reported'],
      dtype='object')
Overlap between numeric and object columns: set()


In [19]:
#lets check if total_claim_amount is equal to the sum of injury_claim, property_claim, and vehicle_claim
df['calculated_total_claim'] = df['injury_claim'] + df['property_claim'] + df['vehicle_claim']
discrepancies = df[df['total_claim_amount'] != df['calculated_total_claim']]
print("Number of discrepancies in total_claim_amount:", len(discrepancies))

Number of discrepancies in total_claim_amount: 0


In [20]:
#So, no missing values that we need to deal with here, or duplicate rows
#We have already dealt with the negative value in the umbrella_limit column
#We have also checked the capital-gains column for negative values and NaN values, and found none
#We have also checked for duplicate rows and found none
#We have also checked for missing values in the dataframe and found none that we need to deal with
#We have also checked the authorities_contacted column for missing values and found none that we need to deal with

#so next step is to save the cleaned dataframe to a new csv file
df.to_csv('insurance_claims_cleaned.csv', index=False)